python/lsst/daf/ingest/index_exposure_task.py

```diff
@@ -0,0 +1,358 @@
+
+    def index(self, exp, data_id, dstype, database):
+        """Spatially index an |exposure| or |metadata| object.
+
+        Parameters
+        ----------
+
+        exp : lsst.afw.image.Exposure[DFILU] or lsst.daf.base.PropertySet
+            An afw |exposure| or corresponding |metadata| object.
+
+        data_id : object
+            An object identifying a single exposure (e.g. as used by the
+            butler). It must be possible to pickle `data_id`.
+
+        dstype : str
+            The dataset type of the exposure.
+
+        database : str
+            A file name identifying the SQLite 3 database to populate.
+        """
+        if isinstance(exp, daf_base.PropertySet):
+            md = exp
+            # Map (LTV1, LTV2) to LSST (x0, y0). LSST convention says that
+            # (x0, y0) is the location of the sub-image origin (the bottom-left
+            # corner) relative to the origin of the parent, whereas LTVi encode
+            # the origin of the parent relative to the origin of the subimage.
+            x0 = -md.get('LTV1') if md.exists('LTV1') else 0
+            y0 = -md.get('LTV2') if md.exists('LTV2') else 0
+            width = md.get('NAXIS1')
+            height = md.get('NAXIS2')
+            wcs = afw_image.makeWcs(md, False)
+        else:
+            x0 = exp.getX0()
+            y0 = exp.getY0()
+            width = exp.getWidth()
+            height = exp.getHeight()
+            wcs = exp.getWcs()
+        # Create a convex polygon containing the exposure pixels. When sphgeom
+        # gains support for non-convex polygons, this could be changed to map
+        # exposure.getPolygon() to a spherical equivalent, or to subdivide box
+        # edges in pixel space to account for non linear projections. This
+        # would have higher accuracy than the current approach of connecting
+        # corner sky coordinates with great circles.
+        corners = [
+            wcs.pixelToSky(x0 - 0.5, y0 - 0.5).toIcrs(),
+            wcs.pixelToSky(x0 - 0.5, y0 + height - 0.5).toIcrs(),
+            wcs.pixelToSky(x0 + width - 0.5, y0 + height - 0.5).toIcrs(),
+            wcs.pixelToSky(x0 + width - 0.5, y0 - 0.5).toIcrs()
+        ]
+        corners = [
+            LonLat.fromRadians(c.getLongitude().asRadians(),
+                               c.getLatitude().asRadians())
+            for c in corners
+        ]
+        poly = ConvexPolygon([UnitVector3d(c) for c in corners])
+        bbox = poly.getBoundingBox3d()
+        # Encode the polygon and dataId into binary strings. Note that in
+        # Python 2, the sqlite3 module maps between Python buffer objects and
+        # BLOBs. When migrating to Python 3, buffer() will have to be changed
+        # to bytes().
+        encoded_polygon = buffer(poly.encode())
+        pickled_data_id = buffer(pickle.dumps(data_id))
+        # Construct quoted table names
+        table_name = quote_sqlite3_identifier(dstype)
```

Comment:
Is it necessary to have separate tables for each dataset type?  Presumably the data I have for `raw` and `calexp` should be close, and if I have the data for `calexp` then I don't care about `raw` and can overwrite it.  I can do the same thing any time I have better data.